### PR TITLE
[BACKLOG-40523] - update ClusterSchemas for projects

### PR DESCRIPTION
**Pull request by abryant-hv**
- refactor Database managers into the new SharedObjectInterface stack
- refactored a few things in Spoon to make them use common code for shared objects
- refactor TransMeta to use local managers instead of a list for ClusterSchemas

### CODE REVIEW

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 1]
```diff
@@ -2662,9 +2665,7 @@ public void newDatabasePartitioningSchema() {
 
   public void newClusteringSchema() {
     TransMeta transMeta = getActiveTransformation();
-    if ( transMeta != null ) {
-      newClusteringSchema( transMeta );
-    }
+    newClusteringSchema( transMeta );
```
[the remark this answers is not in the document]
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2668]: No, so we now check this in the Delegate class instead.

**Comment by abryant-hv** [comment]
> Do we still need Spoon.setDatabaseMenuItems(). Can the Database right click menu items use the TreePopupMenuProvider to provide the right click menu actions like the rest of Shared objects

It probably could, but there were enough subtle differences that I didn't switch it over. Same with the Delegates.

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdesai16) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2668]: Is there a difference between getting the slaveServers from transMeta vs getting from slaveServerManagementInterface.getAll(). 
I have to do something similar for PartitionSchema and want to make sure I am not missing anything

[the remark this answers is not in the document]
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2668]: For *Execution* (including selecting from a list to be referenced by something else) they should always be gotten from the combined list, which is generally the Meta, but could be the Execution Bowl if there is no Meta. 

For duplicate checks, only the current level should be used.

File: ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java [hunk 2]
```diff
@@ -85,27 +104,42 @@ private void showSaveError( ClusterSchema clusterSchema, KettleException e ) {
...
     }
   }
 
-  public void delClusterSchema( TransMeta transMeta, ClusterSchema clusterSchema ) {
+  public void delClusterSchema( ClusterSchemaManagementInterface manager, ClusterSchema clusterSchema ) {
     try {
-
-      int idx = transMeta.getClusterSchemas().indexOf( clusterSchema );
-      transMeta.getClusterSchemas().remove( idx );
+      manager.remove( clusterSchema );
```
Comment (pdesai16) [inline ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java:142]: Is it  prompting for confirmation dialog before deleting

File: ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java [hunk 3]
```diff
@@ -85,27 +104,42 @@ private void showSaveError( ClusterSchema clusterSchema, KettleException e ) {
...
+      // execution bowl for listing all slave servers
+      SlaveServerManagementInterface slaveServerManagementInterface =
+        spoon.getExecutionBowl().getManager( SlaveServerManagementInterface.class );
+
+      ClusterSchemaDialog dialog =
+          new ClusterSchemaDialog( spoon.getShell(), clusterSchema, manager.getAll(),
+                                   slaveServerManagementInterface.getAll() );
+      if ( dialog.open() ) {
+        String newName = clusterSchema.getName().trim();
+        manager.add( clusterSchema );
```
Comment (pdesai16) [inline ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java:119]: Is there a check to make sure there is no deplicate before adding

File: ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java (hunk 2 above)
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java:142]: It does not appear so. This is the existing behavior, and the story doesn't say it should be any different.

File: ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java (hunk 3 above)
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonClustersDelegate.java:119]: The ClusterSchemaDialog handles that.

**Review by pdesai16 — APPROVED** [review]

**Pull request merged**